File: src/main/java/org/moinex/ui/common/FundamentalMetricPaneController.java
/*
 * Filename: FundamentalMetricPaneController.java
 * Created on: January  9, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.common;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import javafx.fxml.FXML;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.scene.text.Text;
import lombok.NoArgsConstructor;
import org.json.JSONObject;
import org.moinex.service.I18nService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Controller;

/**
 * Controller for the Fundamental Metric Pane
 * Displays a single fundamental analysis metric in a card format
 *
 * @note prototype is necessary so that each pane instance is unique
 */
@Controller
@Scope("prototype")
@NoArgsConstructor
public class FundamentalMetricPaneController {
    @FXML private VBox rootVBox;
    @FXML private Text metricNameText;
    @FXML private Label metricValueLabel;
    @FXML private HBox metadataBox;

    private I18nService i18nService;
    private static final BigDecimal THRESHOLD_DECREASE_FONT_SIZE =
            new BigDecimal("1E12"); // 1 Trillion

    @Autowired
    public FundamentalMetricPaneController(I18nService i18nService) {
        this.i18nService = i18nService;
    }

    public Node getRoot() {
        return rootVBox;
    }

    /**
     * Update the metric pane with data
     *
     * @param metricName     Name of the metric
     * @param metricData     JSON object containing metric data
     * @param lastUpdateDate Last update date to use for real-time data without reference_date
     */
    public void updateMetricPane(String metricName, Object metricData, String lastUpdateDate) {
        metricNameText.setText(metricName);

        String valueStr;

        if (metricData instanceof JSONObject) {
            JSONObject metric = (JSONObject) metricData;

            // Set main value
            valueStr = formatMetricValue(metric);
            metricValueLabel.setText(valueStr);

            // Clear and populate metadata
            metadataBox.getChildren().clear();

            // Use reference_date if available, otherwise use lastUpdateDate for real-time data
            String dateToShow = null;
            if (metric.has("reference_date")) {
                dateToShow = metric.getString("reference_date");
            } else if (metric.has("data_temporality")
                    && "real_time".equals(metric.getString("data_temporality"))
                    && lastUpdateDate != null) {
                dateToShow = lastUpdateDate;
            }

            if (dateToShow != null) {
                Label dateLabel =
                        new Label(
                                i18nService.tr(
                                                Constants.TranslationKeys
                                                        .FUNDAMENTAL_ANALYSIS_REFERENCE_DATE)
                                        + ": "
                                        + formatDate(dateToShow));
                dateLabel.getStyleClass().add("metric-metadata-label");
                metadataBox.getChildren().add(dateLabel);
            }
        } else {
            // Simple value
            valueStr = UIUtils.getOrDefault(metricData, Constants.NA_DATA).toString();
            metricValueLabel.setText(valueStr);
            metadataBox.getChildren().clear();
        }

        // Define metric font size based on value
        BigDecimal value = extractNumericValue(metricData);

        if (value.abs().compareTo(THRESHOLD_DECREASE_FONT_SIZE) > 0) {
            metricValueLabel.setStyle("-fx-font-size: 20px;");
        } else {
            metricValueLabel.setStyle("-fx-font-size: 22px;");
        }
    }

    /**
     * Format metric value based on its type
     */
    private String formatMetricValue(JSONObject metric) {
        Object value = metric.opt("value");
        String type = metric.optString("type", "number");

        if (value == null || value.toString().equals("null")) {
            return Constants.NA_DATA;
        }

        try {
            BigDecimal numValue = new BigDecimal(value.toString());

            return switch (type) {
                case "percent" -> UIUtils.formatPercentageForFundamentalAnalysis(numValue);
                case "currency" -> UIUtils.formatCurrency(numValue);
                default -> UIUtils.formatNumWithDecimalPlaces(numValue, 2);
            };
        } catch (Exception e) {
            return UIUtils.getOrDefault(value, Constants.NA_DATA).toString();
        }
    }

    /**
     * Format date string
     */
    private String formatDate(String dateStr) {
        try {
            LocalDateTime date = LocalDateTime.parse(dateStr);
            return UIUtils.formatDateForDisplay(date, i18nService);
        } catch (Exception e) {
            return dateStr;
        }
    }

    private static BigDecimal extractNumericValue(Object metricData) {
        if (metricData == null) {
            return BigDecimal.ZERO;
        }

        if (metricData instanceof JSONObject metric) {
            Object value = metric.opt("value");

            if (value == null || value == JSONObject.NULL) {
                return BigDecimal.ZERO;
            }

            try {
                return new BigDecimal(value.toString());
            } catch (NumberFormatException e) {
                return BigDecimal.ZERO;
            }
        }

        if (metricData instanceof Number) {
            return new BigDecimal(metricData.toString());
        }

        return BigDecimal.ZERO;
    }
}

File: src/main/kotlin/org/moinex/ui/common/FundamentalMetricPaneController.kt
/*
 * Filename: FundamentalMetricPaneController.kt (original filename: FundamentalMetricPaneController.java)
 * Created on: January  9, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 19/03/2026
 */

package org.moinex.ui.common

import javafx.fxml.FXML
import javafx.scene.Node
import javafx.scene.control.Label
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.scene.text.Text
import org.json.JSONObject
import org.moinex.common.constants.Constants
import org.moinex.common.constants.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.service.PreferencesService
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Controller
import java.math.BigDecimal
import java.time.LocalDateTime

@Controller
@Scope("prototype")
class FundamentalMetricPaneController(
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var rootVBox: VBox

    @FXML
    private lateinit var metricNameText: Text

    @FXML
    private lateinit var metricValueLabel: Label

    @FXML
    private lateinit var metadataBox: HBox

    companion object {
        private val THRESHOLD_DECREASE_FONT_SIZE = BigDecimal("1E12")
        private const val LARGE_VALUE_FONT_SIZE = "-fx-font-size: 20px;"
        private const val NORMAL_FONT_SIZE = "-fx-font-size: 22px;"
        private const val METRIC_METADATA_LABEL_STYLE = "metric-metadata-label"
        private const val VALUE_FIELD = "value"
        private const val TYPE_FIELD = "type"
        private const val PERCENT_TYPE = "percent"
        private const val CURRENCY_TYPE = "currency"
        private const val REFERENCE_DATE_FIELD = "reference_date"
        private const val DATA_TEMPORALITY_FIELD = "data_temporality"
        private const val REAL_TIME = "real_time"
    }

    fun getRoot(): Node = rootVBox

    fun updateMetricPane(
        metricName: String,
        metricData: Any?,
        lastUpdateDate: String?,
    ) {
        metricNameText.text = metricName

        when (metricData) {
            is JSONObject -> {
                metricValueLabel.text = formatMetricValue(metricData)
                metadataBox.children.clear()

                val dateToShow = determineDateToShow(metricData, lastUpdateDate)
                dateToShow?.let { addDateLabel(it) }
            }
            else -> {
                metricValueLabel.text = UIUtils.getOrDefault(metricData, Constants.NA_DATA).toString()
                metadataBox.children.clear()
            }
        }

        updateFontSize(extractNumericValue(metricData))
    }

    private fun determineDateToShow(
        metric: JSONObject,
        lastUpdateDate: String?,
    ): String? =
        when {
            metric.has(REFERENCE_DATE_FIELD) -> metric.getString(REFERENCE_DATE_FIELD)
            metric.has(DATA_TEMPORALITY_FIELD) &&
                metric.getString(DATA_TEMPORALITY_FIELD) == REAL_TIME &&
                lastUpdateDate != null -> lastUpdateDate
            else -> null
        }

    private fun addDateLabel(dateStr: String) {
        val dateLabel =
            Label(
                "${preferencesService.translate(
                    TranslationKeys.FUNDAMENTAL_ANALYSIS_REFERENCE_DATE,
                )}: ${formatDate(dateStr)}",
            ).apply {
                styleClass.add(METRIC_METADATA_LABEL_STYLE)
            }
        metadataBox.children.add(dateLabel)
    }

    private fun formatMetricValue(metric: JSONObject): String {
        val value = metric.opt(VALUE_FIELD)
        val type = metric.optString(TYPE_FIELD, "number")

        if (value == null || value.toString() == "null") {
            return Constants.NA_DATA
        }

        return runCatching {
            val numValue = BigDecimal(value.toString())
            when (type) {
                PERCENT_TYPE -> UIUtils.formatPercentageForFundamentalAnalysis(numValue)
                CURRENCY_TYPE -> UIUtils.formatCurrency(numValue)
                else -> UIUtils.formatNumWithDecimalPlaces(numValue, 2)
            }
        }.getOrElse {
            UIUtils.getOrDefault(value, Constants.NA_DATA).toString()
        }
    }

    private fun formatDate(dateStr: String): String =
        runCatching {
            val date = LocalDateTime.parse(dateStr)
            UIUtils.formatDateForDisplay(date)
        }.getOrDefault(dateStr)

    private fun updateFontSize(value: BigDecimal) {
        metricValueLabel.style =
            if (value.abs() > THRESHOLD_DECREASE_FONT_SIZE) {
                LARGE_VALUE_FONT_SIZE
            } else {
                NORMAL_FONT_SIZE
            }
    }

    private fun extractNumericValue(metricData: Any?): BigDecimal =
        when (metricData) {
            null -> BigDecimal.ZERO
            is JSONObject -> {
                val value = metricData.opt(VALUE_FIELD)
                when {
                    value == null || value == JSONObject.NULL -> BigDecimal.ZERO
                    else ->
                        runCatching {
                            BigDecimal(value.toString())
                        }.getOrDefault(BigDecimal.ZERO)
                }
            }
            is Number -> BigDecimal(metricData.toString())
            else -> BigDecimal.ZERO
        }
}
